File: cpp/76_minimum_window_substring.cpp
#include <array>
#include <string>

using namespace std;

class Solution {
 public:
  string minWindow(string s, string t) {
    if (s.size() < t.size()) {
      return "";
    }

    array<int, 58> letter_freq{};
    for (const auto c : t) {
      letter_freq[c - 'A']++;
    }

    int start = 0;
    int min_start = 0, min_len = s.size() + 1;
    int diff = t.size();

    for (int end = 0; end < s.size(); end++) {
      int& freq = letter_freq[s[end] - 'A'];
      if (freq > 0) {
        diff--;
      }
      freq--;

      while (diff == 0) {
        if (end - start + 1 < min_len) {
          min_len = end - start + 1;
          min_start = start;
        }
        int& freq = letter_freq[s[start] - 'A'];
        if (freq >= 0) {
          diff++;
        }
        freq++;
        start++;
      }
    }

    return (min_len < s.size() + 1) ? s.substr(min_start, min_len) : "";
  }
};

File: rust/76_minimum_window_substring.rs
impl Solution {
    pub fn min_window(s: String, t: String) -> String {
        let (s, t) = (s.as_bytes(), t.as_bytes());
        if s.len() < t.len() {
            return "".to_string();
        }

        let mut letter_freq = [0; 58];
        for c in t {
            letter_freq[(c - b'A') as usize] += 1;
        }

        let mut start = 0;
        let (mut min_start, mut min_len) = (0, s.len() + 1);
        let mut diff = t.len();

        for (end, c) in s.iter().enumerate() {
            let freq = &mut letter_freq[(c - b'A') as usize];
            if *freq > 0 {
                diff -= 1;
            }
            *freq -= 1;

            while diff == 0 {
                if end - start + 1 < min_len {
                    min_len = end - start + 1;
                    min_start = start;
                }
                let freq = &mut letter_freq[(s[start] - b'A') as usize];
                if *freq >= 0 {
                    diff += 1;
                }
                *freq += 1;
                start += 1;
            }
        }

        if min_len < s.len() + 1 {
            std::str::from_utf8(&s[min_start..min_start + min_len])
                .unwrap()
                .to_string()
        } else {
            "".to_string()
        }
    }
}
